pkg/informant/endpoints.go

```diff
@@ -3,22 +3,89 @@ package informant
 // This file contains the high-level handlers for various HTTP endpoints
 
 import (
+	"errors"
 	"fmt"
 
 	klog "k8s.io/klog/v2"
 
 	"github.com/neondatabase/autoscaling/pkg/api"
+	"github.com/neondatabase/autoscaling/pkg/util"
 )
 
 // State is the global state of the informant
 type State struct {
 	agents *AgentSet
+	cgroup *CgroupState
 }
 
-// NewState instantiates a new State object
-func NewState(agents *AgentSet) *State {
-	return &State{
+// NewStateOpts are individual options provided to NewState
+type NewStateOpts struct {
+	setFields func(*State)
+	post      func(*State) error
+}
+
+// NewState instantiates a new State object, starting whatever background processes might be
+// required
+//
+// Optional configuration may be provided by NewStateOpts.
+//
+// When WithCgroup is provided, the State object will handle "memory high" events for the cgroup,
+// and
+func NewState(agents *AgentSet, opts ...NewStateOpts) (*State, error) {
+	s := &State{
 		agents: agents,
+		cgroup: nil,
+	}
+	for _, opt := range opts {
+		opt.setFields(s)
+	}
+	for _, opt := range opts {
+		if err := opt.post(s); err != nil {
+			return nil, err
+		}
+	}
+
+	return s, nil
+}
+
+// WithCgroup creates a NewStateOpts that sets its CgroupHandler
+//
+// This function will panic if the provided CgroupConfig is invalid.
+func WithCgroup(cgm *CgroupManager, config CgroupConfig) NewStateOpts {
+	if config.OOMBufferBytes == 0 {
+		panic("invalid CgroupConfig: OOMBufferBytes == 0")
+	} else if config.MaxUpscaleWaitMillis == 0 {
+		panic("invalid CgroupConfig: MaxUpscaleWaitMillis == 0")
+	}
+
+	return NewStateOpts{
+		setFields: func(s *State) {
+			if s.cgroup != nil {
+				panic("WithCgroupHandler option provided more than once")
+			}
+
+			upscaleEventsSendr, upscaleEventsRecvr := util.NewCondChannelPair()
+			s.cgroup = &CgroupState{
+				mgr:                cgm,
+				config:             config,
+				upscaleEventsSendr: upscaleEventsSendr,
+				upscaleEventsRecvr: upscaleEventsRecvr,
+			}
+		},
+		post: func(s *State) error {
+			// FIXME: This is technically racy across restarts. The sequence would be:
+			//  1. Respond "ok" to a downscale request
+			//  2. Restart
+			//  3. Read system memory
+			//  4. Get downscaled (as approved earlier)
+			// A potential way to fix this would be writing to a file to record approved downscale
+			// operations.
+			if err := s.cgroup.setMemoryHigh(); err != nil {
+				return fmt.Errorf("Error setting initial cgroup memory.high: %w", err)
+			}
+			go s.cgroup.handleCgroupSignalsLoop(config)
```

Comment:
one of my "rules" is "if you start a go routine without passing it a context, you should be really sure this is right, and then probably not do it," there are exceptions, but I've always regretted making them, for anything non-trivial.